04-onion-routing.md

```diff
@@ -238,11 +232,44 @@ Field descriptions:
      leaking its position in the route.
 
    * `padding`: This field is for future use and also for ensuring that future non-0-`realm`
-     `per_hop`s won't change the overall `hops_data` size.
+     `hop_data`s won't change the overall `hop_payloads` size.
+
+When forwarding HTLCs, nodes MUST construct the outgoing HTLC as specified
+within `hop_data` above; otherwise, deviation from the specified HTLC
+parameters may lead to extraneous routing failure.
 
-When forwarding HTLCs, nodes MUST construct the outgoing HTLC as specified within
-`per_hop` above; otherwise, deviation from the specified HTLC parameters
-may lead to extraneous routing failure.
+### `tlv_payload` payload format
+
+This is a more flexible format, which avoids the redundant `short_channel_id` field for the final node. 
+
+1. tlv: `tlv_payload`
+2. types:
+    1. type: 2 (`amt_to_forward`)
+    2. data:
+        * [`integer`:`amt_to_forward`]
```

Comment:
What do you mean by `integer` here?
I would have defined it like this:

- [`varint`:`0x02`] (`type`) (`amt_to_forward`)
- [`varint`:`length`]
- [`length`:`amt_to_forward`]

Another note (more about TLV): I think this should be a TLV for a more generic `amount_msat` type that can be used everywhere we need to encode an amount in msat. This is orthogonal to this PR (and should probably be done afterwards) but I think we should move the definition of often-used TLVs to a specific file and reference it from here.

Same for the other TLVs defined below.

Reply:
These really are TLV types, hence the use of `integer` as the shortest possible representation of the numeric value we want to transfer. Indeed this ends up being a short-hand for what you write, however it was copy-pasted from #604 by @rustyrussell. If possible I'd like to keep the shorthand since it gets unwieldy to read otherwise.